src/platform/datapath_linux.c

```diff
@@ -1409,60 +1374,71 @@ CxPlatSocketContextSendComplete(
-            }
-        };
-
-        int Ret =
-            epoll_ctl(
-                ProcContext->EpollFd,
-                EPOLL_CTL_MOD,
-                SocketContext->SocketFd,
-                &SockFdEpEvt);
-        if (Ret != 0) {
-            Status = Ret;
-            QuicTraceEvent(
-                DatapathErrorStatus,
-                "[data][%p] ERROR, %u, %s.",
-                SocketContext->Binding,
-                Status,
-                "epoll_ctl failed");
-            goto Exit;
+    struct epoll_event SockFdEpEvt = {
+        .events = EPOLLIN | EPOLLET,
+        .data = {
+            .ptr = &SocketContext->EventContexts[QUIC_SOCK_EVENT_SOCKET]
         }
+    };
 
-        SocketContext->SendWaiting = FALSE;
+    int Ret =
+        epoll_ctl(
+            ProcContext->EpollFd,
+            EPOLL_CTL_MOD,
+            SocketContext->SocketFd,
+            &SockFdEpEvt);
+    if (Ret != 0) {
+        Status = Ret;
+        QuicTraceEvent(
+            DatapathErrorStatus,
+            "[data][%p] ERROR, %u, %s.",
+            SocketContext->Binding,
+            Status,
+            "epoll_ctl failed");
+        goto Exit;
     }
 
-    while (!CxPlatListIsEmpty(&SocketContext->PendingSendContextHead)) {
-        CXPLAT_SEND_DATA* SendContext =
-            CXPLAT_CONTAINING_RECORD(
-                CxPlatListRemoveHead(&SocketContext->PendingSendContextHead),
-                CXPLAT_SEND_DATA,
-                PendingSendLinkage);
+    CxPlatLockAcquire(&SocketContext->PendingSendContextLock);
+    if (!CxPlatListIsEmpty(&SocketContext->PendingSendContextHead)) {
+        SendContext =
+        CXPLAT_CONTAINING_RECORD(
+            SocketContext->PendingSendContextHead.Flink,
+            CXPLAT_SEND_DATA,
+            PendingSendLinkage);
+    }
+    CxPlatLockRelease(&SocketContext->PendingSendContextLock);
+    if (SendContext == NULL) {
+        goto Exit;
+    }
 
+    do {
         Status =
-            CxPlatSocketSend(
+            CxPlatSocketSendInternal(
                 SocketContext->Binding,
                 SendContext->Bind ? &SendContext->LocalAddress : NULL,
                 &SendContext->RemoteAddress,
-                SendContext);
-        if (QUIC_FAILED(Status)) {
-            goto Exit;
+                SendContext,
+                TRUE);
+        CxPlatLockAcquire(&SocketContext->PendingSendContextLock);
+        if (Status != QUIC_STATUS_PENDING) {
+            CxPlatListRemoveHead(&SocketContext->PendingSendContextHead);
+            CxPlatSendDataFree(SendContext);
         }
-
-        if (SocketContext->SendWaiting) {
-            break;
+        if (!CxPlatListIsEmpty(&SocketContext->PendingSendContextHead)) {
+            SendContext =
+                CXPLAT_CONTAINING_RECORD(
+                    SocketContext->PendingSendContextHead.Flink,
+                    CXPLAT_SEND_DATA,
+                    PendingSendLinkage);
+        } else {
+            SendContext = NULL;
         }
```

Comment:
This should just go inside the previous `if` statement. If `CxPlatSocketSendInternal` returns pending, we should just be bailing out anyways.